stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/experimental/ChatSocket.kt

```diff
@@ -0,0 +1,296 @@
+        networkStateProvider.subscribe(networkStateListener)
+    }
+
+    private suspend fun disposeObservers() {
+        lifecycleObserver.dispose(lifecycleHandler)
+        disposeNetworkStateObserver()
+    }
+
+    private fun disposeNetworkStateObserver() {
+        networkStateProvider.unsubscribe(networkStateListener)
+    }
+
+    private fun handleError(error: ChatError) {
+        logger.e { error.stringify() }
+        when (error) {
+            is ChatNetworkError -> onChatNetworkError(error)
+            else -> callListeners { it.onError(error) }
+        }
+    }
+
+    private fun onChatNetworkError(error: ChatNetworkError) {
+        if (ChatErrorCode.isAuthenticationError(error.streamCode)) {
+            tokenManager.expireToken()
+        }
+
+        when (error.streamCode) {
+            ChatErrorCode.UNDEFINED_TOKEN.code,
+            ChatErrorCode.INVALID_TOKEN.code,
+            ChatErrorCode.API_KEY_NOT_FOUND.code,
+            ChatErrorCode.VALIDATION_ERROR.code,
+            -> {
+                logger.d {
+                    "One unrecoverable error happened. Error: ${error.stringify()}. Error code: ${error.streamCode}"
+                }
+                chatSocketStateService.onUnrecoverableError(error)
+            }
+            else -> chatSocketStateService.onNetworkError(error)
+        }
+    }
+
+    fun removeListener(listener: SocketListener) {
+        synchronized(listeners) {
+            listeners.remove(listener)
+        }
+    }
+
+    fun addListener(listener: SocketListener) {
+        synchronized(listeners) {
+            listeners.add(listener)
+        }
+    }
+
+    /**
+     * Attempt to send [event] to the web socket connection.
+     * Returns true only if socket is connected and [okhttp3.WebSocket.send] returns true, otherwise false
+     *
+     * @see [okhttp3.WebSocket.send]
+     */
+    internal fun sendEvent(event: ChatEvent): Boolean = streamWebSocket?.send(event) ?: false
+
+    internal fun isConnected(): Boolean = chatSocketStateService.currentState is State.Connected
+
+    /**
+     * Get connection id of this connection.
+     */
+    internal fun connectionIdOrError(): String = when (val state = chatSocketStateService.currentState) {
+        is State.Connected -> state.event.connectionId
+        else -> error("This state doesn't contain connectionId")
+    }
+
+    fun reconnectUser(user: User, isAnonymous: Boolean) {
+        chatSocketStateService.onReconnect(
+            when (isAnonymous) {
+                true -> SocketFactory.ConnectionConf.AnonymousConnectionConf(wssUrl, apiKey, user)
+                false -> SocketFactory.ConnectionConf.UserConnectionConf(wssUrl, apiKey, user)
+            }
+        )
+    }
+
+    private fun callListeners(call: (SocketListener) -> Unit) {
+        synchronized(listeners) {
+            listeners.forEach { listener ->
+                coroutineScope.launch(DispatcherProvider.Main) { call(listener) }
```

Comment:
Do we really need to run the listeners asynchronously? This will keep the issue `https://github.com/GetStream/stream-chat-android/issues/4017`. If so, could we keep running asynchronously in the cases where this is a must and run synchronously where possible?

Reply:
Another problem that may happen: 
When `scope.coroutineContext.cancelChildrenExcept(currentJob)` runs in ChatClient, those listeners may be0 canceled prematurely, because there's no guarantee that they will have finished their jobs on time.

Reply:
The issue #4017 is not a problem anymore on this implementation because when `ChatClient.disconnect()` is called, it is the customer who is stoping our SDK, and because we cancel all coroutineScope into our SDk, the coroutineScope that will call the listeners is stopped as well.

Reply:
I did not understand. Won't the cancelled scope avoid that the listeners for disconnection run?